File: examples/java/Ecosystem.java
import io.fluxui.App;
import io.fluxui.Backend;
import io.fluxui.Dsl;
import io.fluxui.FluxUI;
import io.fluxui.Query;
import io.fluxui.Rule;
import io.fluxui.Schema;
import io.fluxui.State;
import io.fluxui.Store;

import java.nio.file.Path;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * FluxUI Ecosystem Demo (Java) — the modern Next.js/Zod/Zustand-style toolkit.
 *
 * Shows the five ecosystem features in one app:
 *   1. Routing + params  — addRoute("/user/:id"), router.param("id"), query()
 *   2. Store<T>          — Zustand-style global state shared across views
 *   3. Schema + Rule     — Zod-style runtime validation
 *   4. Query<T>          — async fetch with IDLE/LOADING/SUCCESS/ERROR + view()
 *   5. Dsl.skeleton      — loading placeholders for a pending Query
 */
public final class Ecosystem {

    /** Global store (Zustand-style): a tiny cart shared by every view. */
    static final class CartState {
        int count;
        String last = "(none)";
    }

    static final Store<CartState> cart = new Store<>(new CartState());

    static Schema signupSchema() {
        return new Schema()
            .field("email", Rule.string().email())
            .field("age", Rule.number().min(18).max(120))
            .field("nick", Rule.string().minLength(3).maxLength(12));
    }

    /** A fake API query that resolves to a fixed payload after a short delay. */
    static final Query<String> users = new Query<>(() -> {
        try {
            Thread.sleep(1200);
        } catch (InterruptedException ignored) {
            Thread.currentThread().interrupt();
        }
        return "Ada Lovelace, Alan Turing, Grace Hopper";
    });

    public static void main(String[] args) {
        if (args.length > 0) {
            FluxUI.loadFrom(Path.of(args[0]));
        }

        try (App app = App.create()) {
            app.setBackend(Backend.AUTO);
            if (!app.init("FluxUI Ecosystem (Java)", 1100, 760)) {
                return;
            }
            app.loadDefaultFont(16.0f);
            app.addStylesheet(
                ".root { display:flex; flex-direction:row; background:#0d1117; color:#e6edf3; }" +
                ".sidebar { display:flex; flex-direction:column; width:220px; background:#161b22; padding:18px; gap:6px; }" +
                ".sidebar button { text-align:left; padding:10px 12px; border-radius:8px; color:#c9d1d9; }" +
                ".sidebar button.active { background:#1f6feb; color:white; }" +
                ".content { display:flex; flex-direction:column; flex-grow:1; padding:28px; gap:18px; }" +
                "h1 { font-size:26px; font-weight:700; }" +
                ".card { background:#161b22; border:1px solid #30363d; border-radius:12px; padding:20px; gap:10px; display:flex; flex-direction:column; }" +
                ".row { display:flex; flex-direction:row; gap:10px; align-items:center; }" +
                ".primary { background:#238636; color:white; padding:10px 18px; border-radius:8px; }" +
                ".ok { color:#3fb950; } .danger { color:#f85149; }" +
                ".skeleton { display:flex; flex-direction:column; gap:10px; }" +
                ".skeleton-line { height:14px; border-radius:6px; background:#21262d; }"
            );

            Dsl.Router router = Dsl.router(app);

            // Route 1 — Store: global state shared between view and sidebar.
            router.addRoute("/store", () -> Dsl.div(
                Dsl.h1("Store (Zustand-style)"),
                Dsl.div(
                    Dsl.p("A single global store, mutated via a reducer."),
                    Dsl.textFn(() -> "Items in cart: " + cart.get().count),
                    Dsl.textFn(() -> "Last added: " + cart.get().last),
                    Dsl.div(
                        Dsl.button("Add apple").className("primary")
                            .onClick(() -> cart.set(s -> { s.count++; s.last = "apple"; })),
                        Dsl.button("Add book").className("primary")
                            .onClick(() -> cart.set(s -> { s.count++; s.last = "book"; })),
                        Dsl.button("Reset")
                            .onClick(() -> cart.set(s -> { s.count = 0; s.last = "(none)"; }))
                    ).className("row")
                ).className("card")
            ));

            // Route 2 — Schema + Rule: Zod-style validation of sample input.
            router.addRoute("/form", () -> {
                State<String> result = new State<>("Press a button to validate a sample payload.");
                return Dsl.div(
                    Dsl.h1("Schema validation (Zod-style)"),
                    Dsl.div(
                        Dsl.textFn(result::get).className("ok"),
                        Dsl.div(
                            Dsl.button("Validate good payload").className("primary").onClick(() -> {
                                Map<String, String> d = new LinkedHashMap<>();
                                d.put("email", "a@b.com");
                                d.put("age", "25");
                                d.put("nick", "joe");
                                Schema.Result res = signupSchema().validate(d);
                                result.set(res.ok ? "Valid! Welcome aboard." : "Invalid: " + res.first());
                            }),
                            Dsl.button("Validate bad payload").onClick(() -> {
                                Map<String, String> d = new LinkedHashMap<>();
                                d.put("email", "nope");
                                d.put("age", "12");
                                d.put("nick", "x");
                                Schema.Result res = signupSchema().validate(d);
                                result.set(res.ok ? "Valid! Welcome aboard." : "Invalid: " + res.first());
                            })
                        ).className("row")
                    ).className("card")
                );
            });

            // Route 3 — Query + Skeleton: async fetch with loading placeholders.
            router.addRoute("/query", () -> {
                if (users.status() == Query.Status.IDLE) {
                    users.start();
                }
                return Dsl.div(
                    Dsl.h1("Query (React-Query-style)"),
                    Dsl.div(
                        Dsl.p("Fetches on a background thread. Shows a Skeleton while loading."),
                        users.view(
                            () -> Dsl.skeleton(3),
                            data -> Dsl.text("Users: " + data).className("ok"),
                            err -> Dsl.text("Error: " + err).className("danger")
                        )
                    ).className("card")
                );
            });

            // Route 4 — Route params + query string: "/user/:id?tab=..."
            router.addRoute("/user/:id", () -> {
                String id = router.param("id");
                String tab = router.query("tab");
                if (tab.isEmpty()) {
                    tab = "profile";
                }
                final String tabFinal = tab;
                return Dsl.div(
                    Dsl.h1("User #" + id),
                    Dsl.div(
                        Dsl.p("Route pattern: /user/:id — captured from the URL path."),
                        Dsl.text("id param  = " + id),
                        Dsl.text("tab query = " + tabFinal),
                        Dsl.div(
                            Dsl.button("Go to user 42").onClick(() -> router.navigate("/user/42?tab=settings")),
                            Dsl.button("Go to user 99").onClick(() -> router.navigate("/user/99?tab=info"))
                        ).className("row")
                    ).className("card")
                );
            });

            // Layout shell: sidebar highlights the active route + a content slot.
            router.setLayout(content -> Dsl.div(
                Dsl.nav(
                    Dsl.el("h2", Dsl.text("FluxUI")),
                    navBtn(router, "Store", "/store"),
                    navBtn(router, "Form (Zod)", "/form"),
                    navBtn(router, "Query", "/query"),
                    navBtn(router, "User #7", "/user/7?tab=info")
                ).className("sidebar"),
                Dsl.div(content).id("__content__").className("content")
            ).className("root"));

            router.build("/store");
            router.runReactive();
        }
    }

    private static io.fluxui.Node navBtn(Dsl.Router router, String label, String path) {
        String cls = router.route().equals(path) ? "active" : "";
        return Dsl.button(label).className(cls).onClick(() -> router.navigate(path));
    }
}

File: examples/zig/ecosystem.zig
// FluxUI Ecosystem Demo (Zig) — the modern Next.js/Zod/Zustand-style toolkit.
//
// Shows the five ecosystem features in one app:
//   1. Routing + params  — addRoute("/user/:id"), d.param("id")
//   2. Store(T)          — Zustand-style global state shared across views
//   3. Schema + Rule     — Zod-style runtime validation
//   4. Query(T)          — fetch with idle/loading/success/err + view()
//   5. skeleton          — loading placeholders for a pending Query
//
// Zig has no closures-with-captures over the C ABI, so views are plain fns and
// state lives in module-level globals (the idiomatic Zig pattern here).
const fluxui = @import("fluxui.zig");
const d = fluxui.dsl;
const std = @import("std");

// ---- Global store (Zustand-style): a tiny cart shared by every view. --------
const CartState = struct { count: i32 = 0, last: [*:0]const u8 = "(none)" };
var cart = d.Store(CartState).init(.{});

// Buffer for the reactive "items in cart" line.
var g_count_buf: [64]u8 = undefined;

fn cartCountText(_: ?*anyopaque) callconv(.c) [*:0]const u8 {
    const s = std.fmt.bufPrintZ(&g_count_buf, "Items in cart: {d}", .{cart.get().count}) catch return "Items in cart: ?";
    return s.ptr;
}

fn addApple(_: ?*fluxui.c.FluxUIWidget, _: ?*anyopaque) callconv(.c) void {
    var st = cart.get();
    st.count += 1;
    st.last = "apple";
    cart.set(st);
}
fn addBook(_: ?*fluxui.c.FluxUIWidget, _: ?*anyopaque) callconv(.c) void {
    var st = cart.get();
    st.count += 1;
    st.last = "book";
    cart.set(st);
}
fn resetCart(_: ?*fluxui.c.FluxUIWidget, _: ?*anyopaque) callconv(.c) void {
    cart.set(.{});
}

// ---- Schema (Zod-style) + a sample validation into a global result. ---------
var g_form_result: [128]u8 = undefined;
var g_form_msg: [*:0]const u8 = "Press a button to validate a sample payload.";

fn signupSchema() d.Schema {
    return d.Schema.init(&.{
        .{ .name = "email", .rule = d.Rule.string().email() },
        .{ .name = "age", .rule = d.Rule.number().min(18).max(120) },
        .{ .name = "nick", .rule = d.Rule.string().minLength(3).maxLength(12) },
    });
}

fn setResult(res: d.ValidationResult) void {
    if (res.ok) {
        g_form_msg = "Valid! Welcome aboard.";
    } else {
        const s = std.fmt.bufPrintZ(&g_form_result, "Invalid: {s}", .{res.first()}) catch {
            g_form_msg = "Invalid.";
            return;
        };
        g_form_msg = s.ptr;
    }
}

fn validateGood(_: ?*fluxui.c.FluxUIWidget, _: ?*anyopaque) callconv(.c) void {
    setResult(signupSchema().validate(&.{
        .{ .name = "email", .value = "a@b.com" },
        .{ .name = "age", .value = "25" },
        .{ .name = "nick", .value = "joe" },
    }));
}
fn validateBad(_: ?*fluxui.c.FluxUIWidget, _: ?*anyopaque) callconv(.c) void {
    setResult(signupSchema().validate(&.{
        .{ .name = "email", .value = "nope" },
        .{ .name = "age", .value = "12" },
        .{ .name = "nick", .value = "x" },
    }));
}
fn formResultText(_: ?*anyopaque) callconv(.c) [*:0]const u8 {
    return g_form_msg;
}

// ---- Query(T): a fetcher that resolves to a fixed payload. -------------------
const UsersQuery = d.Query([*:0]const u8);
var users = UsersQuery.init(fetchUsers);

fn fetchUsers() UsersQuery.QueryError![*:0]const u8 {
    return "Ada Lovelace, Alan Turing, Grace Hopper";
}

var g_users_buf: [128]u8 = undefined;
fn usersLoading() d.Node {
    return d.skeleton(3);
}
fn usersSuccess(data: [*:0]const u8) d.Node {
    const s = std.fmt.bufPrintZ(&g_users_buf, "Users: {s}", .{data}) catch return d.text("Users").class("ok");
    return d.text(s.ptr).class("ok");
}
fn usersError(err: [*:0]const u8) d.Node {
    _ = err;
    return d.text("Error fetching users").class("danger");
}

// ---- Navigation handlers (C-ABI). -------------------------------------------
fn goStore(_: ?*fluxui.c.FluxUIWidget, _: ?*anyopaque) callconv(.c) void {
    d.navigate("/store");
}
fn goForm(_: ?*fluxui.c.FluxUIWidget, _: ?*anyopaque) callconv(.c) void {
    d.navigate("/form");
}
fn goQuery(_: ?*fluxui.c.FluxUIWidget, _: ?*anyopaque) callconv(.c) void {
    d.navigate("/query");
}
fn goUser(_: ?*fluxui.c.FluxUIWidget, _: ?*anyopaque) callconv(.c) void {
    d.navigate("/user/7?tab=info");
}
fn goUser42(_: ?*fluxui.c.FluxUIWidget, _: ?*anyopaque) callconv(.c) void {
    d.navigate("/user/42?tab=settings");
}

// ---- Views. -----------------------------------------------------------------
fn storeView() d.Node {
    return d.div(&.{
        d.h1("Store (Zustand-style)"),
        d.div(&.{
            d.p("A single global store, mutated and read across views."),
            d.textFn(cartCountText, null),
            d.div(&.{
                d.button("Add apple").class("primary").onClick(addApple, null),
                d.button("Add book").class("primary").onClick(addBook, null),
                d.button("Reset").onClick(resetCart, null),
            }).class("row"),
        }).class("card"),
    });
}

fn formView() d.Node {
    return d.div(&.{
        d.h1("Schema validation (Zod-style)"),
        d.div(&.{
            d.textFn(formResultText, null).class("ok"),
            d.div(&.{
                d.button("Validate good payload").class("primary").onClick(validateGood, null),
                d.button("Validate bad payload").onClick(validateBad, null),
            }).class("row"),
        }).class("card"),
    });
}

fn queryView() d.Node {
    if (users.status() == .idle) users.start();
    return d.div(&.{
        d.h1("Query (React-Query-style)"),
        d.div(&.{
            d.p("Runs a fetcher and renders per state. Skeleton on loading."),
            users.view(usersLoading, usersSuccess, usersError),
        }).class("card"),
    });
}

var g_user_title: [64]u8 = undefined;
var g_user_id: [80]u8 = undefined;
fn userView() d.Node {
    const id = d.param("id");
    const title = std.fmt.bufPrintZ(&g_user_title, "User #{s}", .{id}) catch "User";
    const idline = std.fmt.bufPrintZ(&g_user_id, "id param = {s}", .{id}) catch "id param";
    return d.div(&.{
        d.h1(title.ptr),
        d.div(&.{
            d.p("Route pattern: /user/:id — captured from the URL path."),
            d.text(idline.ptr),
            d.div(&.{
                d.button("Go to user 42").onClick(goUser42, null),
            }).class("row"),
        }).class("card"),
    });
}

// ---- App shell. -------------------------------------------------------------
fn layout(content: d.Node) d.Node {
    return d.div(&.{
        d.nav(&.{
            d.el("h2", &.{d.text("FluxUI")}),
            d.button("Store").onClick(goStore, null),
            d.button("Form (Zod)").onClick(goForm, null),
            d.button("Query").onClick(goQuery, null),
            d.button("User #7").onClick(goUser, null),
        }).class("sidebar"),
        d.div(&.{content}).id("__content__").class("content"),
    }).class("root");
}

pub fn main() !void {
    const app = try fluxui.App.create();
    defer app.deinit();

    try app.init("FluxUI Ecosystem (Zig)", 1100, 760);
    _ = app.loadDefaultFont(16.0);

    app.addStylesheet(
        ".root { display:flex; flex-direction:row; background:#0d1117; color:#e6edf3; }" ++
        ".sidebar { display:flex; flex-direction:column; width:220px; background:#161b22; padding:18px; gap:6px; }" ++
        ".sidebar button { text-align:left; padding:10px 12px; border-radius:8px; color:#c9d1d9; }" ++
        ".content { display:flex; flex-direction:column; flex-grow:1; padding:28px; gap:18px; }" ++
        "h1 { font-size:26px; font-weight:700; }" ++
        ".card { background:#161b22; border:1px solid #30363d; border-radius:12px; padding:20px; gap:10px; display:flex; flex-direction:column; }" ++
        ".row { display:flex; flex-direction:row; gap:10px; align-items:center; }" ++
        ".primary { background:#238636; color:white; padding:10px 18px; border-radius:8px; }" ++
        ".ok { color:#3fb950; } .danger { color:#f85149; }" ++
        ".skeleton { display:flex; flex-direction:column; gap:10px; }" ++
        ".skeleton-line { height:14px; border-radius:6px; background:#21262d; }",
    );

    app.addRoute("/store", storeView);
    app.addRoute("/form", formView);
    app.addRoute("/query", queryView);
    app.addRoute("/user/:id", userView);
    app.setLayout(layout);

    try app.build("/store");
    app.runReactive();
}
